### PR TITLE
Add: DWH design conventions

### PR DESCRIPTION
**Pull request by jtcohen6**
@ericalouie happy to collaborate on this more! It probably reads a little too fun/Medium-y right now; I'm both not sorry and very much open to feedback / change requests.

### CODE REVIEW

File: dwh_design_conventions.md [hunk 1]
```diff
@@ -110,14 +110,14 @@ staging:
 
 ### Marts of Facts
```
Comment (ericalouie) [inline dwh_design_conventions.md]: We should add screenshots of how the directories should look. We can use FTA's internal analytics?

File: dwh_design_conventions.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            # salesforce_account_base.sql
+        stg:
+            materialized: view
+            # stg_salesforce_account.sql
+    snowplow:
+        schema: stg_snowplow
+        ...
+```
+
+### Marts of Facts
```
Comment (ericalouie) [inline dwh_design_conventions.md]: We should add screenshots of how the directories should look. We can use FTA's internal analytics?

Comment (jtcohen6) [reply dwh_design_conventions.md]: Yes 100%. Screenshots from Atom or dbt docs?

File: dwh_design_conventions.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+## Fishtown's Thoughts
+
+To that end, we've spent time developing a vocabulary and an architecture for 
+thinking about data warehouse design, working in concert with our dbt-driven 
+modeling [best practices](https://docs.getdbt.com/docs/best-practices). 
+The [coding conventions](https://github.com/fishtown-analytics/corp/blob/master/dbt_coding_conventions.md)
+are still as true as ever; we're applying the same thinking 
+beyond the `select` statement, to the repo and warehouse levels.
+
+### Note on Custom Schemata
```
Comment (jthandy) [inline dwh_design_conventions.md]: this section is a bit wordy--is it actually needed?

File: dwh_design_conventions.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+Depending on the quality of raw data, this base model could also be a staging
+model. A staged view might just as well warrant several models' worth of cleaning, 
+correcting, and categorizing. Raw data comes as it is; staging data is clean, 
+presentable, and ready for the curtain to rise.
+
+Staging models **can** have joins in them to field additional columns for context 
+or enrichment; add rows through unions and remove them through filters;
+deduplicate a natural key or hash together a 
+[surrogate one](https://github.com/fishtown-analytics/dbt-utils#surrogate_key-source); 
+but they should **not** have `group by` aggregations that would change their
```
Comment (jthandy) [inline dwh_design_conventions.md]: *should not perform any operations 

it would also be a bad idea to, for example, join on a dates table. this would change the granularity.

File: dwh_design_conventions.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            # stg_salesforce_account.sql
+    snowplow:
+        schema: stg_snowplow
+        ...
+```
+
+### Marts of Facts
+
+Complex logical rollups can and should happen across several models, with
+discrete steps and abstractions split into queries of 100 lines or fewer. The
```
Comment (jthandy) [inline dwh_design_conventions.md]: i would say "typically 100 lines or fewer". I don't think that's a hard-and-fast rule.

File: dwh_design_conventions.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+end of these rollups is a table, exposed to end-users and aptly prefixed `fct_`:
+it contains the facts that a good reporter seeks.
+
+Marts are stores of one or more fact tables, related primarily by their use case
+in analysis and reporting. They could be organized by business unit, high-level 
+concept, or intended set of end-users: `marketing`, `finance`, `product`, 
+etc. Some fact tables form the basis for other ones and are used across the
+project; you might gird these keystones into a `core` folder, and `ref()` them
+at will.
```
Comment (jthandy) [inline dwh_design_conventions.md]: In traditional Kimball-style warehousing, models like these are filled with so-called "conformed dimensions".

https://www.kimballgroup.com/data-warehouse-business-intelligence-resources/kimball-techniques/dimensional-modeling-techniques/conformed-dimension/

File: dwh_design_conventions.md (hunk 2 above)
Comment (jthandy) [inline dwh_design_conventions.md]: You didn't actually say anything about the theory behind fact models. There is a lot to say here to get people to conceptual what is meant by "fact" in this context. We should take this out of Agile Data Warehouse Design.

File: dwh_design_conventions.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+downstream of fact tables in your DAG and deserves special status.
+
+**Packages:** While not a model folder within your main project, packages
+that include models (like our [snowplow](https://github.com/fishtown-analytics/snowplow) 
+package) can be configured into custom schema and materialization patterns
+from `dbt_project.yml`.
+
+### Postscript
+
+These thoughts are and continue to be in-progress. Fixing the scourge of
```
Comment (jthandy) [inline dwh_design_conventions.md]: This is a bit too flowery :) Try this:

1. Producing code is a collaborative exercise, and the most important goal of this guide is to standardize the way that we work across our team. As such, the bar will be quite high for PRs to get approved that don't follow these conventions.

2. This guide, like all Fishtown Analytics standards, is a work-in-progress. As we continue to build more tooling and improve our consulting practices, we'll update this guide accordingly. If you have suggestions on how to make this guide better, submit a PR.

**Comment by jthandy** [comment]
I think it is not crazy to have parts of this live in discourse! But I don't want that to be the primary destination for this. The goal of this originally was to have an asset that defined how _we_ work in the same way that the style guide does. These are not necessarily suggestions for others. Others can also adopt these practices if they like, but I would like this document to primarily be for an internal audience--consultants at Fishtown.

Happy to discuss further, this conversation has become rather length and might be better handled in meatspace.

**Comment by jtcohen6** [comment]
Coup de gras: https://discourse.getdbt.com/t/how-we-structure-our-dbt-projects/355/1
